### DISCUSSION TITLE
Error in eigsh in task 9

### DISCUSSION THREAD

**Discussion by benpymer** [Part-II-Computational-Physics/.github#41] [General]
Hi!

I'm just trying to finish up task 9. Everything else seems to be working perfectly well at the moment. My code looks like this

upper = 16
lower = 2
Ns2 = np.arange(lower, upper + 1, 1)
h = 1

timingCompactSparse = []
for N in Ns:
    print(f"N={N}")
    result = %timeit -o spsl.eigsh(SparseHamiltonian(N))
    timingCompactSparse.append(result)

timingCompactLinear = []
for N in Ns:
    print(f"N={N}")
    result = %timeit -o spsl.eigsh(LinearHamiltonian(N))
    timingCompactLinear.append(result)

# make the plot
plt.errorbar(Ns, [time.average for time in timingCompact], yerr = [time.stdev for time in timingCompact])
plt.errorbar(Ns2, [time.average for time in timingCompactSparse], yerr = [time.stdev for time in timingCompactSparse])
plt.errorbar(Ns2, [time.average for time in timingCompactLinear], yerr = [time.stdev for time in timingCompactLinear])

# label and rescale the plot
plt.yscale('log')
plt.xlabel('Matrix Size, $N$')
plt.ylabel('Mean Time to Evaluate Eigenvalues / $s$')

I'm using the results from task 4 without re-evaluating them here. I'm getting the error message 

"TypeError: Cannot use scipy.linalg.eigh for sparse A with k >= N. Use scipy.linalg.eigh(A.toarray()) or reduce k." 

however I've no idea what k is or how to reduce it. Does anyone know what's going on? Thanks!

**Comment by JNHallen** [comment]
Hi! 

scipy.linalg.eigsh can only find k eigenvalues and eigenvectors out of the N eigenvalues/eigenvectors that an $N\times N$ matrix actually has. We never care about more than the first one or two eigenstates, so if you change your code to use `spsl.eigsh(SparseHamiltonian(N), k=2)` and `spsl.eigsh(LinearHamiltonian(N), k=2)` that should do it. 

Another useful command to include in eigsh is to set "which='SA'". This ensures that scipy gives you the smallest eigenvalues with corresponding eigenvectors. 

Hope that helps!

**Reply by benpymer** [reply]
Thanks so much! I think that's helped!
